### DISCUSSION TITLE
Switch between A2DP and line input

### DISCUSSION THREAD

**Discussion by SandeepJagdev** [pschatzmann/arduino-audio-tools#503] [Q&A] [answered]
i made two function out of which bluetooth(bt()) function work fine every time but when i switch to line input (rca())there is no output.please help.

void bt()
{
  auto cfg = kit.defaultConfig(TX_MODE);
  cfg.sd_active = false;
  kit.begin(cfg);
  a2dp_sink.set_stream_reader(read_data_stream, false);
  a2dp_sink.start("AudioWave");
 
}
void rca()
{
uint16_t sample_rate=44100;
uint16_t channels = 2;
  a2dp_sink.end();
  auto config = kit.defaultConfig(RXTX_MODE);
  config.sample_rate = sample_rate; 
  config.channels = channels;
  config.sd_active = false;
  config.input_device = AUDIO_HAL_ADC_INPUT_LINE2;
  kit.begin(config);
}

**Comment by pschatzmann** [comment]
Your description does not make any sense to me. You should open I2S for input and output and you write data whenever you want and read data whenever you want. 

There should not be any need for doing any switching whatsoever...

**Reply by SandeepJagdev** [reply]
i make audio player in which i use two inputs 1.bluetooth 2.line input. i want to switch one of them manually

**Reply by pschatzmann** [reply]
Same answer!

**Comment by SandeepJagdev** [comment]
thanks problem solved. i use many input ouput so there is some delay in "copier(). can this work done by other core of esp32. Do it divide RAM and create problem. or any alternative of copier().

**Comment by pschatzmann** [comment]
If you want to decrease any delay you need to decrease the size of all involved buffers: https://github.com/pschatzmann/arduino-audio-tools/wiki/Performance

**Comment by SandeepJagdev** [comment]
thanks done , one last problem is that equalizer work on a2dp but not in line input

**Comment by pschatzmann** [comment]
If this is the case you did not structure the flow right: Just put it before I2S and make sure that all output is written to the Equilizer

**Comment by SandeepJagdev** [comment]
what is alternative of command

 a2dp_sink.set_stream_reader(read_data_stream, false);

for line input.

**Reply by pschatzmann** [reply]
Why would you need an alternative?

**Comment by SandeepJagdev** [comment]
in example code if i write copier to equalizer then there is error. i do not understand how i can write incoming data to equalizer.

#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"

AudioKitStream kit; // Access I2S as stream
StreamCopy copier(kit, kit); // copy kit to kit
ConfigEquilizer3Bands cfg_eq;
// Arduino Setup

void read_data_stream(const uint8_t *data, uint32_t length) {
    eq.write(data, length);
}

void setup(void) {
    Serial.begin(115200);
    AudioLogger::instance().begin(Serial, AudioLogger::Warning);
    
    auto cfg = kit.defaultConfig(RXTX_MODE);
    cfg.sd_active = false;
    cfg.input_device = AUDIO_HAL_ADC_INPUT_LINE2;
    kit.begin(cfg);
 cfg_eq = eq.defaultConfig();
  cfg_eq.setAudioInfo(cfg); // use channels, bits_per_sample and sample_rate from kit
  cfg_eq.gain_low = 0.5; 
  cfg_eq.gain_medium = 0.5;
  cfg_eq.gain_high = 1.0;
  eq.begin(cfg_eq);

  // register callback
  **************.set_stream_reader(read_data_stream, false);//////////////////////////how can i write to data stream
}

// Arduino loop - copy data
void loop() {
   copier.copy();
}

**Comment by pschatzmann** [comment, marked as answer]
I dont see the Equalizer defined anywhere in your code but the copy should rather be StreamCopy copier(eq, kit); // copy kit to eq
if you want to send the output thru the Equilizer...

**Comment by SandeepJagdev** [comment]
thank you very much.
